core/modules/wdb/ChunkResource.cc

```diff
@@ -257,46 +263,44 @@ class ChunkEntry {
-            SubChunkMap& scm = ti->second;
-            SubChunkMap::iterator si, se;
-            for(si=scm.begin(), se=scm.end(); si != se; ++si) {
-                if (si->second == 0) {
-                    discardable.push_back(ScTable(db, _chunkId,
-                                                  ti->first,
-                                                  si->first));
-                    mapDiscardable.push_back(si->first);
-                } else if (si->second < 0) {
-                    throw Bug("ChunkResource ChunkEntry::flush: Invalid negative use count when flushing subchunks");
-                }
-            } // All subchunks
-            // Prune zero elements for this db+table+chunk
-            // (invalidates iterators)
-            IntVector::iterator di, de;
-            for(di=mapDiscardable.begin(), de=mapDiscardable.end();
-                di != de; ++di) {
-                scm.erase(*di);
-            }
-        } // All tables
-        // Delegate actual table dropping to the backend.
-        if (discardable.size() > 0) {
-            backend->discard(discardable);
-        }
-    }
+         ScTableVector discardable;
+         std::lock_guard<std::mutex> lock(_mutex);
+         backend->memLockRequireOwnership();
+         for(auto& elem : _tableMap) {
+             IntVector mapDiscardable;
+             SubChunkMap& scm = elem.second;
+             SubChunkMap::iterator si, se;
+             for(si=scm.begin(), se=scm.end(); si != se; ++si) {
+                 if (si->second == 0) {
+                     discardable.push_back(ScTable(_chunkId, elem.first, si->first));
+                     mapDiscardable.push_back(si->first);
+                 } else if (si->second < 0) {
+                     throw Bug("ChunkResource ChunkEntry::flush: Invalid negative use count when flushing subchunks");
+                 }
+             } // All subchunks
+             // Prune zero elements for this db+table+chunk
+             // (invalidates iterators)
+             IntVector::iterator di, de;
+             for(di=mapDiscardable.begin(), de=mapDiscardable.end();
+                 di != de; ++di) {
+                 scm.erase(*di);
+             }
+         } // All tables
+         // Delegate actual table dropping to the backend.
+         if (discardable.size() > 0) {
+             backend->discard(discardable);
+         }
+     }
+
+
 private:
     void _release(ScTableVector const& needed) {
         // _mutex should be held.
         // Release subChunkId for the right table
-        for(ScTableVector::const_iterator i=needed.begin(), e=needed.end();
-            i != e; ++i) {
-            SubChunkMap& scm = _tableMap[i->table];
-            int last = scm[i->subChunkId];
-            scm[i->subChunkId] = last - 1;
+        for(auto const& elem : needed) {
+            SubChunkMap& scm = _tableMap[elem.dbTable];
+            int last = scm[elem.subChunkId];
+            scm[elem.subChunkId] = last - 1;
```

Comment:
That would be better... I'm really hoping that's an artifact or the original code.